pilot/pkg/bootstrap/namespacecontroller.go

```diff
@@ -39,27 +40,70 @@ const (
 	namespaceResyncPeriod = time.Second * 30
 	// The name of the ConfigMap in each namespace storing the root cert of non-Kube CA.
-	CACertNamespaceConfigMap      = "istio-ca-root-cert"
-	CACertNamespaceInsertInterval = time.Second
-	CACertNamespaceInsertTimeout  = time.Second * 2
+	CACertNamespaceConfigMap = "istio-ca-root-cert"
 )
 
-// NamespaceController manages the CA certificate in each namespace.
+var (
+	configMapLabel = map[string]string{"istio.io/config": "true"}
+)
+
+// NamespaceController manages reconciles a configmap in each namespace with a desired set of data.
 type NamespaceController struct {
-	ca   *ca.IstioCA
+	// data is the function to fetch the data we will insert into the config map
+	data func() map[string]string
 	core corev1.CoreV1Interface
 
+	queue queue.Instance
+
 	// Controller and store for namespace objects
 	namespaceController cache.Controller
-	namespaceStore      cache.Store
+	// Controller and store for ConfigMap objects
+	configMapController cache.Controller
 }
 
-// NewNamespaceController returns a pointer to a newly constructed SecretController instance.
-func NewNamespaceController(ca *ca.IstioCA, core corev1.CoreV1Interface) (*NamespaceController, error) {
+// NewNamespaceController returns a pointer to a newly constructed NamespaceController instance.
+func NewNamespaceController(data func() map[string]string, core corev1.CoreV1Interface) (*NamespaceController, error) {
 	c := &NamespaceController{
-		ca:   ca,
-		core: core,
+		data:  data,
+		core:  core,
+		queue: queue.NewQueue(time.Second),
 	}
+	configMapLw := listwatch.MultiNamespaceListerWatcher([]string{metav1.NamespaceAll}, func(namespace string) cache.ListerWatcher {
+		return &cache.ListWatch{
+			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
+				options.LabelSelector = fields.SelectorFromSet(configMapLabel).String()
+				return core.ConfigMaps(namespace).List(options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				options.LabelSelector = fields.SelectorFromSet(configMapLabel).String()
+				return core.ConfigMaps(namespace).Watch(options)
+			}}
+	})
+	_, c.configMapController =
+		cache.NewInformer(configMapLw, &v1.ConfigMap{}, 0, cache.ResourceEventHandlerFuncs{
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				c.queue.Push(func() error {
+					return c.configMapChange(newObj)
+				})
+			},
+			DeleteFunc: func(obj interface{}) {
+				cm, ok := obj.(*v1.ConfigMap)
+				if ok {
+					c.queue.Push(func() error {
+						ns, err := core.Namespaces().Get(cm.Namespace, metav1.GetOptions{})
+						if err != nil {
+							return err
+						}
+						// If the namespace is terminating, we may get into a loop of trying to re-add the configmap back
+						// We should make sure the namespace still exists
+						if ns.Status.Phase != v1.NamespaceTerminating {
+							return c.writeToNamespace(cm.Namespace)
+						}
+						return nil
+					})
```

Comment:
I'm not 100% sure if we need to handle `cache.DeletedFinalStateUnknown` here for missed watch events. k8s watch should be reliable. Might be worth adding for robustness? See https://github.com/kubernetes/sample-controller/blob/81e6fe3b90edfeb1e629c3c224214c3002d22c89/controller.go#L355-L367 for an example.